src/utils/BigInt.ts

```diff
@@ -0,0 +1,223 @@
+import Decimal, { type Numeric } from 'decimal.js-light'
+
+Decimal.default.set({
+  precision: 30,
+  toExpNeg: -7,
+  toExpPos: 29,
+  rounding: Decimal.default.ROUND_HALF_CEIL, // ROUND_HALF_CEIL is 1
+})
+
+export function Dec(value: Numeric) {
+  return new Decimal.default(value)
+}
+
+export abstract class BigIntWrapper {
+  protected value: bigint
+
+  constructor(value: Numeric | bigint) {
+    if (typeof value === 'bigint') {
+      this.value = value
+    } else if (value instanceof Decimal.default) {
+      this.value = BigInt(value.toFixed(0))
+    } else if (typeof value === 'number') {
+      this.value = BigInt(Math.floor(value))
+    } else {
+      this.value = BigInt(String(value).split('.')[0] || '0')
+    }
+  }
+
+  toString() {
+    return this.value.toString()
+  }
+
+  toBigInt() {
+    return this.value
+  }
+}
+
+export class DecimalWrapper extends BigIntWrapper {
+  protected decimals: number
+
+  constructor(value: Numeric | bigint, decimals: number) {
+    super(value)
+    this.decimals = decimals
+  }
+
+  static _fromFloat<T extends DecimalWrapper>(num: Numeric, decimals: number) {
+    const n = Dec(num.toString()).mul(Dec(10).pow(decimals)).toDecimalPlaces(0).toString()
+    return new (this as any)(n, decimals) as T
+  }
+
+  toDecimal() {
+    return Dec(this.value.toString()).div(Dec(10).pow(this.decimals))
+  }
+
+  toFloat() {
+    return this.toDecimal().toNumber()
+  }
+
+  _add<T>(value: bigint | (T extends BigIntWrapper ? T : never)): T {
+    const val = typeof value === 'bigint' ? value : value.toBigInt()
+    return new (this.constructor as any)(this.value + val, this.decimals)
+  }
+
+  _sub<T>(value: bigint | (T extends BigIntWrapper ? T : never)): T {
+    const val = typeof value === 'bigint' ? value : value.toBigInt()
+    return this._add<T>(-val)
+  }
+
+  _mul<T>(value: bigint | (T extends BigIntWrapper ? T : never)): T {
```

Comment:
I think these will only be useful if they operate on the decimal representation. We may want something like `currency = tokens.mul(price)`. I think it should be multiplying by e.g. `1.1234` and not by `1.1234e18`

Reply:
Good point, I added a case to `Currency.mul()`, if a `Price` is passed to the multiply function it will divide by the Prices decimals and then multiply. We could optionally also expose the `_mulPrice` function to make it more clear